src/neuroglancer/datasource/graphene/frontend.ts

```diff
@@ -0,0 +1,820 @@
+import {DataType} from 'neuroglancer/util/data_type';
+import {Owned} from 'neuroglancer/util/disposable';
+import {mat4, vec3} from 'neuroglancer/util/geom';
+import {HttpError, isNotFoundError, responseJson} from 'neuroglancer/util/http_request';
+import {parseArray, parseFixedLengthArray, verifyEnumString, verifyFiniteFloat, verifyFinitePositiveFloat, verifyInt, verifyObject, verifyObjectProperty, verifyOptionalObjectProperty, verifyOptionalString, verifyPositiveInt, verifyString, verifyNonnegativeInt} from 'neuroglancer/util/json';
+import {getObjectId} from 'neuroglancer/util/object_id';
+import {parseSpecialUrl, SpecialProtocolCredentials, SpecialProtocolCredentialsProvider} from 'neuroglancer/util/special_protocol_request';
+import {Uint64} from 'neuroglancer/util/uint64';
+import {cancellableFetchSpecialOk, getGrapheneFragmentKey, isBaseSegmentId, responseIdentity} from 'neuroglancer/datasource/graphene/base';
+import {ChunkedGraphSourceParameters, MeshSourceParameters, MultiscaleMeshMetadata, PYCG_APP_VERSION, VolumeChunkEncoding} from 'neuroglancer/datasource/graphene/base';
+import {DataEncoding, ShardingHashFunction, ShardingParameters} from 'neuroglancer/datasource/precomputed/base';
+import {ChunkedGraphChunkSource, ChunkedGraphLayer} from 'neuroglancer/sliceview/chunked_graph/frontend';
+import {StatusMessage} from 'neuroglancer/status';
+import { makeChunkedGraphChunkSpecification } from 'neuroglancer/sliceview/chunked_graph/base';
+import { Uint64Set } from 'neuroglancer/uint64_set';
+import { ComputedSplit, SegmentationGraphSource, SegmentationGraphSourceConnection, VISIBLE_SEGMENT_TYPE } from 'neuroglancer/segmentation_graph/source';
+import { VisibleSegmentsState } from 'neuroglancer/segmentation_display_state/base';
+import { WatchableValueInterface } from 'neuroglancer/trackable_value';
+import { RenderLayerTransformOrError } from 'neuroglancer/render_coordinate_transform';
+import { RenderLayer } from 'neuroglancer/renderlayer';
+import { getSegmentPropertyMap, parseProviderUrl, PrecomputedDataSource, PrecomputedVolumeChunkSource } from 'neuroglancer/datasource/precomputed/frontend';
+
+class GrapheneChunkedGraphChunkSource extends
+(WithParameters(WithCredentialsProvider<SpecialProtocolCredentials>()(ChunkedGraphChunkSource), ChunkedGraphSourceParameters)) {}
+
+class GrapheneMeshSource extends
+(WithParameters(WithCredentialsProvider<SpecialProtocolCredentials>()(MeshSource), MeshSourceParameters)) {
+  getFragmentKey(objectKey: string|null, fragmentId: string) {
+    objectKey;
+    return getGrapheneFragmentKey(fragmentId);
+  }
+}
+
+function resolvePath(a: string, b: string) {
+  const outputParts = a.split('/');
+  for (const part of b.split('/')) {
+    if (part === '..') {
+      if (outputParts.length !== 0) {
+        outputParts.length = outputParts.length - 1;
+        continue;
+      }
+    }
+    outputParts.push(part);
+  }
+  return outputParts.join('/');
+}
+
+class ScaleInfo {
```

Comment:
It would be nice to refactor this to share code with the precomputed data source rather than duplicating it, if possible.

Reply:
I pushed these changes
https://github.com/google/neuroglancer/pull/383/commits/3c0028054bedd55afb89d20c5a83b09be8d56a88
https://github.com/google/neuroglancer/pull/383/commits/24ad0ddb6d15d2a71dd50169de8b844debfdfa0f

I think I could re-use PrecomputedDataSource.get though I need to override `getVolumeDataSource`. One way to do that would be to make it a method on PrecomputedDataSource though if that made sense, it would mean moving all of the other getDataSource functions into that class.